extensions/khr/GLSL_KHR_cooperative_matrix.txt

```diff
@@ -0,0 +1,396 @@
+type can be used as.
+
+    const int gl_MatrixUseA             = 0;
+    const int gl_MatrixUseB             = 1;
+    const int gl_MatrixUseAccumulator   = 2;
+
+Add a new Section 5.4.X, Cooperative Matrix Type Constructors
+
+Cooperative matrices can be constructed from a single scalar value whose
+type matches the matrix's component type (or any value that can be
+implicitly converted to that type). This initializes all components of the
+matrix to that same value.
+
+Cooperative matrices can be constructed from another cooperative matrix
+type with the same scope, number of rows, number of columns, and use.
+This performs a component-wise type conversion to initialize the new cooperative matrix.
+
+Add a new Section 5.X, Cooperative Matrix Components
+
+The components of a cooperative matrix are spread across the invocations
+in its scope, in an implementation-dependent manner. The components owned
+by a given invocation can be accessed using array subscripting syntax,
+and the number of components owned by each invocation can be queried
+using the *length* method. The type returned by *length* is an int.
+There is no compile-time bounds checking of array indices.
+
+This can be used, for example, to perform component-wise operations on
+all components of a cooperative matrix:
+
+coopmat<float16_t, gl_ScopeSubgroup, 16, 8, gl_MatrixUseAccumulator> m;
+...
+for (int i = 0; i < m.length(); ++i) {
+	m[i] = f(m[i]);
+}
+
+Modify Section 5.9, Expressions
+
+The arithmetic binary operators add (+), subtract (-), multiply (*), and
+divide (/) operate on cooperative matrix types and perform the operation
+component-wise. The operands must have identical types.
+
+The arithmetic binary operator multiply (*) also operates on a cooperative
+matrix type and a scalar (in either order) and performs the multiply
+component-wise. The scalar type must match the component type of the
+matrix.
+
+The arithmetic unary operator negate (-) operates on cooperative matrix
+types and performs the operation component-wise.
+
+Conversions are allowed between cooperative matrix types assuming the
+element types, scope, row size, column size, and use are the same.
+
+Add a new Section 8.X, Cooperative Matrix Functions
+
+The following functions are used to load and store cooperative matrix
+values from and to memory. In memory, the matrices are stored as arrays
+of their element type and size. In the following functions, the generic type
+"coopmat" can accept an coopmat type with any type parameters. The
+"buf" arrays must be in either buffer storage or shared storage, and the
+array that is passed in can be sized or unsized.
+
+For all of these functions, for a given dynamic instance of the function
+call, all function parameters must be the same for all invocations in a
+given scope instance (where the scope is the scope the cooperative matrix
+type(s) were created with). All invocations in a given scope instance must
+be active or all must be inactive.
```

Comment:
This currently only applies to `coopMatLoad`, `coopMatStore` & `coopMatMulAdd`.  Do we need to say something similar in the "Expressions" section above?

Reply:
A similar question came up at https://github.com/KhronosGroup/SPIRV-Registry/issues/208, see my response there. I'm not planning to make a change in this spec unless we decide something different there.